File: integration_tests/devices/eps.py
import struct
from Queue import Queue, Empty
import logging

from datetime import datetime

import i2cMock
from i2cMock import I2CDevice
from threading import Event

from utils import call, CompareAsDict, bitlist_to_byte


class MPPT(CompareAsDict):
    def __init__(self):
        self.SOL_VOLT = 0
        self.SOL_CURR = 0
        self.SOL_OUT_VOLT = 0
        self.TEMP = 0
        self.STATE = 0

    def bytes(self):
        return list(struct.pack('<HHHHB', self.SOL_CURR, self.SOL_VOLT, self.SOL_OUT_VOLT, self.TEMP, self.STATE))


class DISTR(CompareAsDict):
    def __init__(self):
        self.VOLT_3V3 = 0
        self.CURR_3V3 = 0
        self.VOLT_5V = 0
        self.CURR_5V = 0
        self.VOLT_VBAT = 0
        self.CURR_VBAT = 0
        self.LCL_STATE = 0
        self.LCL_FLAGB = 0

    def bytes(self):
        return list(struct.pack('<HHHHHHBB',
                                self.CURR_3V3,
                                self.VOLT_3V3,
                                self.CURR_5V,
                                self.VOLT_5V,
                                self.CURR_VBAT,
                                self.VOLT_VBAT,
                                self.LCL_STATE,
                                self.LCL_FLAGB
                                ))


class BATC_A(CompareAsDict):
    def __init__(self):
        self.VOLT_A = 0
        self.CHRG_CURR = 0
        self.DCHRG_CURR = 0
        self.TEMP = 0
        self.STATE = 0

    def bytes(self):
        return list(struct.pack('<HHHHB',
                                self.VOLT_A,
                                self.CHRG_CURR,
                                self.DCHRG_CURR,
                                self.TEMP,
                                self.STATE))


class BATC_B(CompareAsDict):
    def __init__(self):
        self.VOLT_B = 0

    def bytes(self):
        return list(struct.pack('<H', self.VOLT_B))


class BP_A(CompareAsDict):
    def __init__(self):
        self.TEMP_A = 0
        self.TEMP_B = 0

    def bytes(self):
        return list(struct.pack('<HH', self.TEMP_A, self.TEMP_B))


class BP_B(CompareAsDict):
    def __init__(self):
        self.TEMP_C = 0

    def bytes(self):
        return list(struct.pack('<H', self.TEMP_C))


class OtherController(CompareAsDict):
    def __init__(self):
        self.VOLT_3V3d = 0

    def bytes(self):
        return list(struct.pack('<H', self.VOLT_3V3d))


class ThisController(CompareAsDict):
    def __init__(self):
        self.SAFETY_CTR = 0
        self.PWR_CYCLES = 0
        self.UPTIME = 0
        self.TEMP = 0
        self.SUPP_TEMP = 0

    def bytes(self):
        return list(struct.pack('<BHIHH',
                                self.SAFETY_CTR,
                                self.PWR_CYCLES,
                                self.UPTIME,
                                self.TEMP,
                                self.SUPP_TEMP))


class DCDC(CompareAsDict):
    def __init__(self):
        self.TEMP = 0

    def bytes(self):
        return list(struct.pack('<H', self.TEMP))


class HousekeepingA(CompareAsDict):
    def __init__(self):
        self.MPPT_X = MPPT()
        self.MPPT_Y_PLUS = MPPT()
        self.MPPT_Y_MINUS = MPPT()
        self.DISTR = DISTR()
        self.BATC = BATC_A()
        self.BP = BP_A()
        self.CTRLB = OtherController()
        self.CTRLA = ThisController()
        self.DCDC3V3 = DCDC()
        self.DCDC5V = DCDC()

    def bytes(self):
        return self.MPPT_X.bytes() \
               + self.MPPT_Y_PLUS.bytes() \
               + self.MPPT_Y_MINUS.bytes() \
               + self.DISTR.bytes() \
               + self.BATC.bytes() \
               + self.BP.bytes() \
               + self.CTRLB.bytes() \
               + self.CTRLA.bytes() \
               + self.DCDC3V3.bytes() \
               + self.DCDC5V.bytes()


class HousekeepingB(CompareAsDict):
    def __init__(self):
        self.BP = BP_B()
        self.BATC = BATC_B()
        self.CTRLA = OtherController()
        self.CTRLB = ThisController()

    def bytes(self):
        return sum(map(lambda x: x.bytes(), [self.BP, self.BATC, self.CTRLA, self.CTRLB]), [])


class LCLTimeoutException(Exception):
    def __init__(self, desired):
        self.desired = desired

    def __str__(self):
        return "Timeout waiting for LCL to become %s" % (self.desired,)


class LCL:
    def __init__(self):
        self.on_event = Event()
        self.off_event = Event()
        self.changes = Queue()
        self.is_on = False
        self.was_on = False
        self.on_enable = None
        self.on_disable = None

    def arrange_on(self):
        self.is_on = True

    def on(self):
        self.is_on = True
        self.was_on = True

        self.on_event.set()
        self.off_event.clear()

        self.changes.put(True)
        call(self.on_enable, True)

    def off(self):
        self.is_on = False

        self.on_event.clear()
        self.off_event.set()

        self.changes.put(False)
        call(self.on_disable, True)

    def wait_for_change(self, timeout=None):
        try:
            return self.changes.get(True, timeout)
        except Empty:
            raise LCLTimeoutException("x")

    def wait_for_on(self, timeout=None):
        if not self.on_event.wait(timeout):
            raise LCLTimeoutException("on")

    def wait_for_off(self, timeout=None):
        if not self.off_event.wait(timeout):
            raise LCLTimeoutException("off")


class BurnSwitch:
    def __init__(self):
        self.enabled = False
        self.on_enable = None

    def enable(self):
        self.enabled = True
        call(self.on_enable, True)


class EPSControllerA(I2CDevice):
    def __init__(self, eps):
        super(EPSControllerA, self).__init__(0x35, "EPS.A")
        self._eps = eps

        self._lcls = [eps.TKmain, eps.SunS, eps.CamNadir, eps.CamWing, eps.SENS, eps.ANTenna]
        self._burn_switches = [eps.SAILmain, eps.SADSmain]

        self.on_power_cycle = None
        self.on_disable_overheat_submode = None
        self.on_get_housekeeping = None
        self.on_reset_watchdog = None

        self.hk = HousekeepingA()

        self.last_watchdog_kick = None

    @i2cMock.command([0xE0])
    def _power_cycle(self):
        self.log.info("Triggered power cycle")
        call(self.on_power_cycle, None)

    @i2cMock.command([0xE1])
    def _enable_lcl(self, lcl_id):
        self.log.info("Enable LCL(%d)", lcl_id)
        self._lcls[lcl_id - 1].on()

    @i2cMock.command([0xE2])
    def _disable_lcl(self, lcl_id):
        self.log.info("Disable LCL(%d)", lcl_id)
        self._lcls[lcl_id - 1].off()

    @i2cMock.command([0xE3])
    def _enable_burn_switch(self, switch_id):
        self.log.info("Enable BURN switch(%d)", switch_id)
        self._burn_switches[switch_id - 1].enable()

    @i2cMock.command([0xE4])
    def _disable_overheat_submode(self):
        self.log.info("Disable overheat submode")
        call(self.on_disable_overheat_submode, None)

    @i2cMock.command([0xE5])
    def _reset_watchdog(self):
        self.log.info("Resetting watchdog")
        self.last_watchdog_kick = datetime.now()
        call(self.on_reset_watchdog, None)

    @i2cMock.command([0x0])
    def _housekeeping(self):
        self.hk.DISTR.LCL_STATE = bitlist_to_byte(map(lambda lcl: 1 if lcl.is_on else 0, self._lcls))
        self.hk.DISTR.LCL_FLAGB = bitlist_to_byte(map(lambda lcl: 1 if lcl.is_on else 0, self._lcls))

        hk = call(self.on_get_housekeeping, default=self.hk)
        return [0x61] + hk.bytes()

    @i2cMock.command([0x4B])
    def _error_code(self):
        hk = call(self.on_get_housekeeping, default=self.hk)
        return hk.bytes()[0x4B - 1:]


class EPSControllerB(I2CDevice):
    def __init__(self, eps):
        super(EPSControllerB, self).__init__(0x36, "EPS.B")
        self._eps = eps

        self._lcls = [eps.TKred, eps.ANTennaRed]
        self._burn_switches = [eps.SAILred, eps.SADSred]

        self.on_power_cycle = None
        self.on_disable_overheat_submode = None
        self.on_get_housekeeping = None
        self.on_reset_watchdog = None

        self.hk = HousekeepingB()

        self.last_watchdog_kick = None

    @i2cMock.command([0xE0])
    def _power_cycle(self):
        self.log.info("Triggered power cycle")
        call(self.on_power_cycle, None)

    @i2cMock.command([0xE1])
    def _enable_lcl(self, lcl_id):
        self.log.info("Enable LCL(%d)", lcl_id)
        self._lcls[lcl_id - 1].on()

    @i2cMock.command([0xE2])
    def _disable_lcl(self, lcl_id):
        self.log.info("Disable LCL(%d)", lcl_id)
        self._lcls[lcl_id - 1].off()

    @i2cMock.command([0xE3])
    def _enable_burn_switch(self, switch_id):
        self.log.info("Enable BURN switch(%d)", switch_id)
        self._burn_switches[switch_id - 1].enable()

    @i2cMock.command([0xE4])
    def _disable_overheat_submode(self):
        self.log.info("Disable overheat submode")
        call(self.on_disable_overheat_submode, None)

    @i2cMock.command([0xE5])
    def _reset_watchdog(self):
        self.log.info("Resetting watchdog")
        self.last_watchdog_kick = datetime.now()
        call(self.on_reset_watchdog, None)

    @i2cMock.command([0x0])
    def _housekeeping(self):
        hk = call(self.on_get_housekeeping, default=self.hk)
        return [0x9D] + hk.bytes()

    @i2cMock.command([0x07])
    def _error_code(self):
        hk = call(self.on_get_housekeeping, default=self.hk)
        return hk.bytes()[0x07 - 1:]


class EPS:
    ERROR_COUNTER = 1

    def __init__(self):
        self.TKmain = LCL()
        self.SunS = LCL()
        self.CamNadir = LCL()
        self.CamWing = LCL()
        self.SENS = LCL()
        self.ANTenna = LCL()
        self.TKred = LCL()
        self.ANTennaRed = LCL()

        self.SAILmain = BurnSwitch()
        self.SAILred = BurnSwitch()
        self.SADSmain = BurnSwitch()
        self.SADSred = BurnSwitch()

        self.controller_a = EPSControllerA(self)
        self.controller_b = EPSControllerB(self)

File: src/commands/eps.cpp
#include "eps/eps.h"
#include <cstdlib>
#include <cstring>
#include "obc_access.hpp"
#include "terminal/terminal.h"

using devices::eps::LCL;
using devices::eps::BurnSwitch;
using devices::eps::EPSDriver;

static void Usage()
{
    GetTerminal().Puts("eps enable_lcl|disable_lcl|power_cycle|disable_overheat|enable_burn_switch|hk_a|hk_b");
}

static void Print(const char* prefix, const devices::eps::hk::MPPT_HK& hk)
{
    GetTerminal().Printf("%s.SOL_VOLT\t%d\n", prefix, hk.SOL_VOLT.Value());
    GetTerminal().Printf("%s.SOL_CURR\t%d\n", prefix, hk.SOL_CURR.Value());
    GetTerminal().Printf("%s.SOL_OUT_VOLT\t%d\n", prefix, hk.SOL_OUT_VOLT.Value());
    GetTerminal().Printf("%s.TEMP\t%d\n", prefix, hk.Temperature.Value());
    GetTerminal().Printf("%s.STATE\t%d\n", prefix, num(hk.MpptState));
}

static void Print(const char* prefix, const devices::eps::hk::DISTR_HK& hk)
{
    GetTerminal().Printf("%s.CURR_3V3\t%d\n", prefix, hk.CURR_3V3.Value());
    GetTerminal().Printf("%s.VOLT_3V3\t%d\n", prefix, hk.VOLT_3V3.Value());
    GetTerminal().Printf("%s.CURR_5V\t%d\n", prefix, hk.CURR_5V.Value());
    GetTerminal().Printf("%s.VOLT_5V\t%d\n", prefix, hk.VOLT_5V.Value());
    GetTerminal().Printf("%s.CURR_VBAT\t%d\n", prefix, hk.CURR_VBAT.Value());
    GetTerminal().Printf("%s.VOLT_VBAT\t%d\n", prefix, hk.VOLT_VBAT.Value());
    GetTerminal().Printf("%s.LCL_STATE\t%d\n", prefix, num(hk.LCL_STATE));
    GetTerminal().Printf("%s.LCL_FLAGB\t%d\n", prefix, num(hk.LCL_FLAGB));
}

static void Print(const char* prefix, const devices::eps::hk::BATCPrimaryState& hk)
{
    GetTerminal().Printf("%s.VOLT_A\t%d\n", prefix, hk.VOLT_A.Value());
    GetTerminal().Printf("%s.CHRG_CURR\t%d\n", prefix, hk.ChargeCurrent.Value());
    GetTerminal().Printf("%s.DCHRG_CURR\t%d\n", prefix, hk.DischargeCurrent.Value());
    GetTerminal().Printf("%s.TEMP\t%d\n", prefix, hk.Temperature.Value());
    GetTerminal().Printf("%s.STATE\t%d\n", prefix, num(hk.State));
}

static void Print(const char* prefix, const devices::eps::hk::BATCSecondaryState& hk)
{
    GetTerminal().Printf("%s.VOLT_B\t%d\n", prefix, hk.voltB.Value());
}

static void Print(const char* prefix, const devices::eps::hk::BatteryPackPrimaryState hk)
{
    GetTerminal().Printf("%s.TEMP_A\t%d\n", prefix, hk.temperatureA.Value());
    GetTerminal().Printf("%s.TEMP_B\t%d\n", prefix, hk.temperatureB.Value());
}

static void Print(const char* prefix, const devices::eps::hk::BatteryPackSecondaryState& hk)
{
    GetTerminal().Printf("%s.TEMP_C\t%d\n", prefix, hk.temperatureC.Value());
}

static void Print(const char* prefix, const devices::eps::hk::OtherControllerState& hk)
{
    GetTerminal().Printf("%s.VOLT_3V3d\t%d\n", prefix, hk.VOLT_3V3d.Value());
}

static void Print(const char* prefix, const devices::eps::hk::ThisControllerState& hk)
{
    GetTerminal().Printf("%s.SAFETY_CTR\t%d\n", prefix, hk.safetyCounter);
    GetTerminal().Printf("%s.PWR_CYCLES\t%d\n", prefix, hk.powerCycleCount);
    GetTerminal().Printf("%s.UPTIME\t%ld\n", prefix, hk.uptime);
    GetTerminal().Printf("%s.TEMP\t%d\n", prefix, hk.temperature.Value());
    GetTerminal().Printf("%s.SUPP_TEMP\t%d\n", prefix, hk.suppTemp.Value());
}

static void Print(const char* prefix, devices::eps::hk::DCDC_HK& hk)
{
    GetTerminal().Printf("%s.TEMP\t%d\n", prefix, hk.temperature.Value());
}

void EPSCommand(std::uint16_t argc, char* argv[])
{
    if (argc == 0)
    {
        Usage();
        return;
    }

    auto& eps = GetEPS();

    if (strcmp(argv[0], "enable_lcl") == 0)
    {
        auto lcl = static_cast<LCL>(atoi(argv[1]));
        auto result = eps.EnableLCL(lcl);

        GetTerminal().Printf("%d", num(result));
        return;
    }

    if (strcmp(argv[0], "disable_lcl") == 0)
    {
        auto lcl = static_cast<LCL>(atoi(argv[1]));
        auto result = eps.DisableLCL(lcl);

        GetTerminal().Printf("%d", num(result));
        return;
    }

    if (strcmp(argv[0], "power_cycle") == 0)
    {
        bool result = false;

        if (strcmp(argv[1], "A") == 0)
        {
            result = eps.PowerCycle(EPSDriver::Controller::A);
        }

        else if (strcmp(argv[1], "B") == 0)
        {
            result = eps.PowerCycle(EPSDriver::Controller::B);
        }
        else if (strcmp(argv[1], "Auto") == 0)
        {
            result = eps.PowerCycle();
        }

        GetTerminal().Puts(result ? "1" : "0");
        return;
    }

    if (strcmp(argv[0], "disable_overheat") == 0)
    {
        bool result = false;

        if (strcmp(argv[1], "A") == 0)
        {
            result = eps.DisableOverheatSubmode(EPSDriver::Controller::A);
        }

        else if (strcmp(argv[1], "B") == 0)
        {
            result = eps.DisableOverheatSubmode(EPSDriver::Controller::B);
        }

        GetTerminal().Puts(result ? "1" : "0");
        return;
    }

    if (strcmp(argv[0], "enable_burn_switch") == 0)
    {
        bool useMain = strcmp(argv[1], "1") == 0;

        auto burnSwitch = static_cast<BurnSwitch>(atoi(argv[2]));

        auto result = eps.EnableBurnSwitch(useMain, burnSwitch);

        GetTerminal().Printf("%d", num(result));
        return;
    }

    if (strcmp(argv[0], "hk_a") == 0)
    {
        auto hk = eps.ReadHousekeepingA();
        if (!hk.HasValue)
        {
            GetTerminal().Puts("Error");
            return;
        }

        auto v = hk.Value;

        Print("MPPT_X", v.mpptX);
        Print("MPPT_Y_PLUS", v.mpptYPlus);
        Print("MPPT_Y_MINUS", v.mpptYMinus);
        Print("DISTR", v.distr);
        Print("BATC", v.batc);
        Print("BP", v.bp);
        Print("CTRLB", v.other);
        Print("CTRLA", v.current);
        Print("DCDC3V3", v.dcdc3V3);
        Print("DCDC5V", v.dcdc5V);

        return;
    }

    if (strcmp(argv[0], "hk_b") == 0)
    {
        auto hk = eps.ReadHousekeepingB();
        if (!hk.HasValue)
        {
            GetTerminal().Puts("Error");
            return;
        }

        auto v = hk.Value;

        Print("BP", v.bp);
        Print("BATC", v.batc);
        Print("CTRLA", v.other);
        Print("CTRLB", v.current);

        return;
    }

    if (strcmp(argv[0], "reset_watchdog") == 0)
    {
        bool result = false;

        if (strcmp(argv[1], "A") == 0)
        {
            result = eps.ResetWatchdog(EPSDriver::Controller::A) == devices::eps::ErrorCode::NoError;
        }

        else if (strcmp(argv[1], "B") == 0)
        {
            result = eps.ResetWatchdog(EPSDriver::Controller::B) == devices::eps::ErrorCode::NoError;
        }
        else if (strcmp(argv[1], "Both") == 0)
        {
            result = eps.ResetWatchdog(EPSDriver::Controller::A) == devices::eps::ErrorCode::NoError;
            result = result && eps.ResetWatchdog(EPSDriver::Controller::B) == devices::eps::ErrorCode::NoError;
        }

        GetTerminal().Puts(result ? "1" : "0");
        return;
    }
}
